analyzers/src/SonarAnalyzer.CSharp/Rules/GenericTypeParameterUnused.cs

```diff
@@ -37,123 +37,82 @@ public sealed class GenericTypeParameterUnused : SonarDiagnosticAnalyzer
-
-                MethodDeclarationSyntax methodDeclaration when IsMethodCandidate(methodDeclaration, semanticModel)
-                    => new ParametersInfo(methodDeclaration.TypeParameterList, "method"),
-
-                var wrapper when LocalFunctionStatementSyntaxWrapper.IsInstance(wrapper)
-                    => new ParametersInfo(((LocalFunctionStatementSyntaxWrapper)node).TypeParameterList, "local function"),
-
-                var wrapper when RecordDeclarationSyntaxWrapper.IsInstance(wrapper)
-                    => new ParametersInfo(((RecordDeclarationSyntaxWrapper)node).TypeParameterList, "record"),
-
-                _ => default
+                StructDeclarationSyntax structDeclaration => new ParametersInfo(structDeclaration.TypeParameterList, "struct"),
+                MethodDeclarationSyntax methodDeclaration when IsMethodCandidate(methodDeclaration, semanticModel) => new ParametersInfo(methodDeclaration.TypeParameterList, "method"),
+                var wrapper when LocalFunctionStatementSyntaxWrapper.IsInstance(wrapper) => new ParametersInfo(((LocalFunctionStatementSyntaxWrapper)node).TypeParameterList, "local function"),
+                var wrapper when RecordDeclarationSyntaxWrapper.IsInstance(wrapper) => new ParametersInfo(((RecordDeclarationSyntaxWrapper)node).TypeParameterList, "record"),
+                _ => null
             };
 
-        private readonly struct ParametersInfo
-        {
-            public TypeParameterListSyntax Parameters { get; }
-
-            public string ContainerName { get; }
-
-            public ParametersInfo(TypeParameterListSyntax parameters, string name)
-            {
-                Parameters = parameters;
-                ContainerName = name;
-            }
-        }
-
-        private static bool IsMethodCandidate(MethodDeclarationSyntax methodDeclaration, SemanticModel semanticModel)
-        {
-            var syntaxValid = !methodDeclaration.Modifiers.Any(modifier => MethodModifiersToSkip.Contains(modifier.Kind()))
-                              && methodDeclaration.ExplicitInterfaceSpecifier == null
-                              && methodDeclaration.HasBodyOrExpressionBody();
-
-            if (!syntaxValid)
-            {
-                return false;
-            }
-
-            var methodSymbol = semanticModel.GetDeclaredSymbol(methodDeclaration);
-
-            return methodSymbol != null && methodSymbol.IsChangeable();
-        }
+        private static bool IsMethodCandidate(MethodDeclarationSyntax methodDeclaration, SemanticModel semanticModel) =>
+            !methodDeclaration.Modifiers.Any(x => MethodModifiersToSkip.Contains(x.Kind()))
+            && methodDeclaration.ExplicitInterfaceSpecifier is null
+            && methodDeclaration.HasBodyOrExpressionBody()
+            && semanticModel.GetDeclaredSymbol(methodDeclaration) is { } methodSymbol
```

Comment:
SemanticModel is bound to a SyntaxTree (a file). `symbol.DeclaringSyntaxReferences` can return a node from another SyntaxTree for partial classes. So we need `EnsureCorrectSemanticModelOrDefault` to transition from a known semantic model to the correct one, belonging to that node.

We need it in `GetUsedTypeParameters`, because this method processes result of `symbol.DeclaringSyntaxReferences`.

We don't need it in `IsMethodCandidate`, because this one processes result of `SyntaxNodeAnalysisContext` items `Node` and `SemanticModel`, so it is by-design guaranteed that it is the correct semantic model.